File: tui/lavadome/handeval.py
"""Poker hand evaluation — the best 5-card hand from 2 to 7 cards.

A port of ``adenosine/packages/cards/src/hand-eval.ts``, which is the source of
truth the Wii port's README names and which itself says *"From Texas Hold'Em
Lava Dome"* — this game is where it came from originally.

It is ported line for line rather than rewritten, because a poker evaluator
that is subtly wrong plays *almost* right, which is worse than one that plainly
does not. ``tests/test_handeval.py`` checks it against the real JavaScript over
thousands of random hands (see ``tools/`` and the README), which is the same
method the web build's ``config.js`` records using when this evaluator replaced
the game's original one.

**The caller supplies ``value``, and it is never rewritten here.** That is what
lets one evaluator serve ace-low and ace-high games. Poker is ace-high, so
cards must come from :mod:`lavadome.cards`, which stamps ace-high at
construction.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from itertools import combinations
from typing import Any

HAND_RANKS = {
    "Royal Flush": 9,
    "Straight Flush": 8,
    "Four of a Kind": 7,
    "Full House": 6,
    "Flush": 5,
    "Straight": 4,
    "Three of a Kind": 3,
    "Two Pair": 2,
    "One Pair": 1,
    "High Card": 0,
}

HAND_POINTS = {
    "Royal Flush": 1000,
    "Straight Flush": 500,
    "Four of a Kind": 250,
    "Full House": 150,
    "Flush": 100,
    "Straight": 75,
    "Three of a Kind": 50,
    "Two Pair": 25,
    "One Pair": 10,
    "High Card": 0,
}


@dataclass
class HandResult:
    """The outcome of evaluating a hand.

    ``points`` is what the dome compares against its threshold, and ``name`` is
    what the payout table is keyed on — those two fields are the game. The rest
    is for display and for breaking ties between candidate five-card hands.
    """

    name: str
    rank: int
    points: int
    tiebreakers: list[int] = field(default_factory=list)
    cards: list[Any] = field(default_factory=list)
    description: str = ""
    partial: bool = False


def _rank_name_for(sorted_cards: list[Any], value: int) -> str:
    for card in sorted_cards:
        if card.value == value:
            return card.rank
    return sorted_cards[0].rank if sorted_cards else ""


class HandEvaluator:
    """Evaluates the best five-card hand out of the cards it is given."""

    def evaluate(self, cards: list[Any]) -> HandResult:
        if not cards or len(cards) < 2:
            return self._empty_result()
        if len(cards) < 5:
            return self._evaluate_partial(cards)

        best: HandResult | None = None
        for combo in combinations(cards, 5):
            result = self._evaluate_five(list(combo))
            # Strictly greater, so the first candidate wins a tie — same as the
            # JS, which matters only for which cards come back, never the name.
            if best is None or self._compare(result, best) > 0:
                best = result
        return best if best is not None else self._empty_result()

    # ── Five-card evaluation ────────────────────────────────────────

    def _evaluate_five(self, cards: list[Any]) -> HandResult:
        ordered = sorted(cards, key=lambda c: c.value, reverse=True)
        is_flush = self._is_flush(cards)
        is_straight = self._is_straight(ordered)
        counts = self._value_counts(cards)
        count_vals = sorted(counts.values(), reverse=True)
        first = count_vals[0] if count_vals else 0
        second = count_vals[1] if len(count_vals) > 1 else 0

        if is_flush and is_straight and ordered[0].value == 14 and ordered[1].value == 13:
            name = "Royal Flush"
            tiebreakers = [14]
        elif is_flush and is_straight:
            name = "Straight Flush"
            tiebreakers = [self._straight_high_card(ordered)]
        elif first == 4:
            name = "Four of a Kind"
            tiebreakers = self._tiebreak_by_count(counts, [4, 1])
        elif first == 3 and second == 2:
            name = "Full House"
            tiebreakers = self._tiebreak_by_count(counts, [3, 2])
        elif is_flush:
            name = "Flush"
            tiebreakers = [c.value for c in ordered]
        elif is_straight:
            name = "Straight"
            tiebreakers = [self._straight_high_card(ordered)]
        elif first == 3:
            name = "Three of a Kind"
            tiebreakers = self._tiebreak_by_count(counts, [3, 1, 1])
        elif first == 2 and second == 2:
            name = "Two Pair"
            tiebreakers = self._tiebreak_by_count(counts, [2, 2, 1])
        elif first == 2:
            name = "One Pair"
            tiebreakers = self._tiebreak_by_count(counts, [2, 1, 1, 1])
        else:
            name = "High Card"
            tiebreakers = [c.value for c in ordered]

        return HandResult(
            name=name,
            rank=HAND_RANKS[name],
            points=HAND_POINTS[name],
            tiebreakers=tiebreakers,
            cards=ordered,
            description=self._describe(name, ordered),
        )

    def _evaluate_partial(self, cards: list[Any]) -> HandResult:
        """Name a two-to-four card hand, for showing progress before the river.

        The order of these checks is the JS order, not a tidied one: Two Pair is
        tested before Full House and Four of a Kind. With fewer than five cards
        the later branches cannot both match, so it comes to the same thing —
        but reordering it would be a change in behaviour on a hand of five, and
        this function is not the one that guards against being called with one.
        """
        ordered = sorted(cards, key=lambda c: c.value, reverse=True)
        counts = self._value_counts(cards)
        count_vals = sorted(counts.values(), reverse=True)
        first = count_vals[0] if count_vals else 0
        second = count_vals[1] if len(count_vals) > 1 else 0

        name = "High Card"
        if first == 2 and second == 2:
            name = "Two Pair"
        elif first == 3 and second == 2:
            name = "Full House"
        elif first == 4:
            name = "Four of a Kind"
        elif first == 3:
            name = "Three of a Kind"
        elif first == 2:
            name = "One Pair"

        return HandResult(
            name=name,
            rank=HAND_RANKS[name],
            points=HAND_POINTS[name],
            tiebreakers=[c.value for c in ordered],
            cards=ordered,
            description=f"{name} (partial)",
            partial=True,
        )

    @staticmethod
    def _empty_result() -> HandResult:
        return HandResult(
            name="No Cards",
            rank=-1,
            points=0,
            tiebreakers=[],
            cards=[],
            description="No cards dealt",
            partial=True,
        )

    # ── Comparison ──────────────────────────────────────────────────

    @staticmethod
    def _compare(a: HandResult, b: HandResult) -> int:
        if a.rank != b.rank:
            return a.rank - b.rank
        for i in range(max(len(a.tiebreakers), len(b.tiebreakers))):
            av = a.tiebreakers[i] if i < len(a.tiebreakers) else 0
            bv = b.tiebreakers[i] if i < len(b.tiebreakers) else 0
            if av != bv:
                return av - bv
        return 0

    # ── Shapes ──────────────────────────────────────────────────────

    @staticmethod
    def _is_flush(cards: list[Any]) -> bool:
        suit = cards[0].suit
        return all(c.suit == suit for c in cards)

    @staticmethod
    def _is_straight(ordered: list[Any]) -> bool:
        """Five in sequence, counting the wheel.

        A-2-3-4-5 is a straight even though the ace sorts to the front, which is
        the one case a plain descending-run check misses.
        """
        if all(ordered[i].value - ordered[i + 1].value == 1
               for i in range(len(ordered) - 1)):
            return True

        values = sorted(c.value for c in ordered)
        return (len(values) == 5 and values[4] == 14
                and values[0] == 2 and values[1] == 3
                and values[2] == 4 and values[3] == 5)

    @staticmethod
    def _straight_high_card(ordered: list[Any]) -> int:
        """The rank a straight is named for.

        In the wheel the ace plays low, so the hand is five high even though the
        ace sorts to the front.
        """
        values = sorted(c.value for c in ordered)
        if (len(values) == 5 and values[4] == 14
                and values[0] == 2 and values[3] == 5):
            return 5
        return ordered[0].value

    def _straight_rank_name(self, ordered: list[Any]) -> str:
        return _rank_name_for(ordered, self._straight_high_card(ordered))

    @staticmethod
    def _value_counts(cards: list[Any]) -> dict[int, int]:
        counts: dict[int, int] = {}
        for card in cards:
            counts[card.value] = counts.get(card.value, 0) + 1
        return counts

    @staticmethod
    def _tiebreak_by_count(counts: dict[int, int], pattern: list[int]) -> list[int]:
        """Rank values ordered by how many of each there are.

        ``pattern`` is the shape being described — ``[3, 2]`` for a full house
        means "the tripled rank, then the paired one" — so a full house of
        threes over kings beats one of twos over aces.
        """
        groups: dict[int, list[int]] = {}
        for value, count in counts.items():
            groups.setdefault(count, []).append(value)
        for group in groups.values():
            group.sort(reverse=True)

        result: list[int] = []
        seen: set[int] = set()
        for target in pattern:
            for value in groups.get(target, ()):
                if value not in seen:
                    result.append(value)
                    seen.add(value)
                    break
        return result

    # ── Description ─────────────────────────────────────────────────

    def _describe(self, name: str, ordered: list[Any]) -> str:
        top = ordered[0]
        if name == "Royal Flush":
            return f"Royal Flush — {top.suit}"
        if name == "Straight Flush":
            return f"Straight Flush — {self._straight_rank_name(ordered)} high"
        if name == "Four of a Kind":
            return f"Four {top.rank}s"
        if name == "Full House":
            counts = self._value_counts(ordered)
            # sorted() for the same reason as Two Pair below. With five cards
            # there is only ever one of each, so this is belt and braces.
            triple = next((v for v, c in sorted(counts.items()) if c == 3), None)
            pair = next((v for v, c in sorted(counts.items()) if c == 2), None)
            if triple is None or pair is None:
                return "Full House"
            return (f"Full House — {_rank_name_for(ordered, triple)}s full of "
                    f"{_rank_name_for(ordered, pair)}s")
        if name == "Flush":
            return f"Flush — {top.rank} high ({top.suit})"
        if name == "Straight":
            return f"Straight — {self._straight_rank_name(ordered)} high"
        if name == "Three of a Kind":
            return f"Three {top.rank}s"
        if name == "Two Pair":
            counts = self._value_counts(ordered)
            # sorted(): a JS object with integer-like keys enumerates in
            # ascending numeric order no matter how it was built, where a
            # Python dict keeps insertion order — which here is descending
            # card order. Without this, kings and fives read "Ks and 5s"
            # where every other port says "5s and Ks".
            pairs = "s and ".join(
                _rank_name_for(ordered, v)
                for v, c in sorted(counts.items()) if c == 2
            )
            return f"Two Pair — {pairs}s"
        if name == "One Pair":
            counts = self._value_counts(ordered)
            pair = next((v for v, c in sorted(counts.items()) if c == 2), None)
            return f"Pair of {_rank_name_for(ordered, pair) if pair else ''}s"
        if name == "High Card":
            return f"{top.rank} high"
        return name

File: wii/source/hand_eval.c
#include <string.h>
#include "hand_eval.h"

/* Points per class, straight out of the web build's HAND_POINTS. The dome's
   threshold starts at 10 and climbs 5 a round, so round 1 is beaten by exactly
   One Pair and round 15 needs a Flush. Change a number here and you have
   changed the difficulty curve, not the wording. */
static const int POINTS[HAND_CLASS_COUNT] = {
    [HAND_HIGH_CARD]       = 0,
    [HAND_ONE_PAIR]        = 10,
    [HAND_TWO_PAIR]        = 25,
    [HAND_THREE_OF_A_KIND] = 50,
    [HAND_STRAIGHT]        = 75,
    [HAND_FLUSH]           = 100,
    [HAND_FULL_HOUSE]      = 150,
    [HAND_FOUR_OF_A_KIND]  = 250,
    [HAND_STRAIGHT_FLUSH]  = 500,
    [HAND_ROYAL_FLUSH]     = 1000
};

static const char *NAMES[HAND_CLASS_COUNT] = {
    "High Card", "One Pair", "Two Pair", "Three of a Kind", "Straight",
    "Flush", "Full House", "Four of a Kind", "Straight Flush", "Royal Flush"
};

int hand_points(HandRank rank) {
    if (rank < 0 || rank >= HAND_CLASS_COUNT) return 0;
    return POINTS[rank];
}

const char *hand_name(HandRank rank) {
    if (rank < 0 || rank >= HAND_CLASS_COUNT) return "No Hand";
    return NAMES[rank];
}

/* Descending, so tiebreakers read most-significant first everywhere below. */
static void sort_desc(int *v, int n) {
    for (int i = 1; i < n; i++) {
        int key = v[i], j = i - 1;
        while (j >= 0 && v[j] < key) {
            v[j + 1] = v[j];
            j--;
        }
        v[j + 1] = key;
    }
}

static int is_flush(const Card *c) {
    for (int i = 1; i < 5; i++) {
        if (c[i].suit != c[0].suit) return 0;
    }
    return 1;
}

/* `desc` is the five ranks, already sorted descending. Returns the rank the
   straight is named by, or 0 when it is not one.
 *
 * The wheel is the whole reason this returns a value rather than a flag. In
 * A-2-3-4-5 the ace plays low, so the hand is five-high, not ace-high -- get
 * that wrong and the wheel silently outranks every other straight. */
static int straight_high(const int *desc) {
    int run = 1;
    for (int i = 0; i < 4; i++) {
        if (desc[i] - desc[i + 1] != 1) { run = 0; break; }
    }
    if (run) return desc[0];

    if (desc[0] == RANK_ACE && desc[1] == 5 && desc[2] == 4 &&
        desc[3] == 3 && desc[4] == 2) {
        return 5;
    }
    return 0;
}

HandResult hand_eval_five(const Card *cards) {
    HandResult r;
    memset(&r, 0, sizeof(r));
    r.valid = 1;

    int desc[5];
    for (int i = 0; i < 5; i++) desc[i] = cards[i].rank;
    sort_desc(desc, 5);

    /* Ranks paired with how many of each, ordered by count then by rank, which
       is exactly the order the tiebreakers want: the trips before the pair, the
       quad before its kicker. */
    int rank_of[5], count_of[5], groups = 0;
    for (int i = 0; i < 5; i++) {
        int found = 0;
        for (int g = 0; g < groups; g++) {
            if (rank_of[g] == desc[i]) { count_of[g]++; found = 1; break; }
        }
        if (!found) {
            rank_of[groups] = desc[i];
            count_of[groups] = 1;
            groups++;
        }
    }
    for (int i = 1; i < groups; i++) {
        int kr = rank_of[i], kc = count_of[i], j = i - 1;
        while (j >= 0 && (count_of[j] < kc ||
                         (count_of[j] == kc && rank_of[j] < kr))) {
            rank_of[j + 1] = rank_of[j];
            count_of[j + 1] = count_of[j];
            j--;
        }
        rank_of[j + 1] = kr;
        count_of[j + 1] = kc;
    }

    int flush = is_flush(cards);
    int high  = straight_high(desc);
    int top   = count_of[0];
    int second = groups > 1 ? count_of[1] : 0;

    if (flush && high == RANK_ACE) {
        r.rank = HAND_ROYAL_FLUSH;
        r.tiebreak[0] = RANK_ACE;
        r.tiebreak_count = 1;
    } else if (flush && high) {
        r.rank = HAND_STRAIGHT_FLUSH;
        r.tiebreak[0] = high;
        r.tiebreak_count = 1;
    } else if (top == 4) {
        r.rank = HAND_FOUR_OF_A_KIND;
        r.tiebreak[0] = rank_of[0];
        r.tiebreak[1] = rank_of[1];
        r.tiebreak_count = 2;
    } else if (top == 3 && second == 2) {
        r.rank = HAND_FULL_HOUSE;
        r.tiebreak[0] = rank_of[0];
        r.tiebreak[1] = rank_of[1];
        r.tiebreak_count = 2;
    } else if (flush) {
        r.rank = HAND_FLUSH;
        for (int i = 0; i < 5; i++) r.tiebreak[i] = desc[i];
        r.tiebreak_count = 5;
    } else if (high) {
        r.rank = HAND_STRAIGHT;
        r.tiebreak[0] = high;
        r.tiebreak_count = 1;
    } else {
        /* Everything left is decided by the grouped order alone: trips, two
           pair, one pair and high card all read their groups most-significant
           first, which the sort above already arranged. */
        if      (top == 3) r.rank = HAND_THREE_OF_A_KIND;
        else if (top == 2 && second == 2) r.rank = HAND_TWO_PAIR;
        else if (top == 2) r.rank = HAND_ONE_PAIR;
        else               r.rank = HAND_HIGH_CARD;

        for (int i = 0; i < groups && i < HAND_TIEBREAK_MAX; i++) {
            r.tiebreak[i] = rank_of[i];
        }
        r.tiebreak_count = groups < HAND_TIEBREAK_MAX ? groups : HAND_TIEBREAK_MAX;
    }

    r.points = POINTS[r.rank];
    return r;
}

int hand_compare(const HandResult *a, const HandResult *b) {
    if (a->rank != b->rank) return a->rank < b->rank ? -1 : 1;

    int n = a->tiebreak_count < b->tiebreak_count
          ? a->tiebreak_count : b->tiebreak_count;
    for (int i = 0; i < n; i++) {
        if (a->tiebreak[i] != b->tiebreak[i]) {
            return a->tiebreak[i] < b->tiebreak[i] ? -1 : 1;
        }
    }
    return 0;
}

HandResult hand_eval_partial(const Card *cards, int count) {
    HandResult r;
    memset(&r, 0, sizeof(r));

    if (count < 2) return r;   /* invalid: nothing to say yet */
    r.valid = 1;

    int desc[HAND_MAX_CARDS];
    for (int i = 0; i < count; i++) desc[i] = cards[i].rank;
    sort_desc(desc, count);

    int best_count = 1, pairs = 0;
    for (int i = 0; i < count; i++) {
        int n = 0;
        for (int j = 0; j < count; j++) if (desc[j] == desc[i]) n++;
        if (n > best_count) best_count = n;
    }
    for (int i = 0; i < count; i++) {
        int n = 0;
        for (int j = 0; j < count; j++) if (desc[j] == desc[i]) n++;
        if (n == 2 && (i == 0 || desc[i] != desc[i - 1])) pairs++;
    }

    if      (best_count >= 4) r.rank = HAND_FOUR_OF_A_KIND;
    else if (best_count == 3 && pairs >= 1) r.rank = HAND_FULL_HOUSE;
    else if (best_count == 3) r.rank = HAND_THREE_OF_A_KIND;
    else if (pairs >= 2)      r.rank = HAND_TWO_PAIR;
    else if (pairs == 1)      r.rank = HAND_ONE_PAIR;
    else                      r.rank = HAND_HIGH_CARD;

    for (int i = 0; i < count && i < HAND_TIEBREAK_MAX; i++) r.tiebreak[i] = desc[i];
    r.tiebreak_count = count < HAND_TIEBREAK_MAX ? count : HAND_TIEBREAK_MAX;

    /* Zero, always. A hand that is not finished has not beaten the dome, and a
       partial result carrying real points is one careless call away from
       resolving a round early. */
    r.points = 0;
    return r;
}

HandResult hand_eval(const Card *cards, int count) {
    HandResult best;
    memset(&best, 0, sizeof(best));

    if (count < 5 || count > HAND_MAX_CARDS) return best;   /* invalid */

    /* Every five-card subset. At seven cards that is 21 of them, which is
       cheaper than any cleverness would be and impossible to get subtly wrong. */
    int idx[5];
    int have_best = 0;

    for (idx[0] = 0;        idx[0] < count - 4; idx[0]++)
    for (idx[1] = idx[0]+1; idx[1] < count - 3; idx[1]++)
    for (idx[2] = idx[1]+1; idx[2] < count - 2; idx[2]++)
    for (idx[3] = idx[2]+1; idx[3] < count - 1; idx[3]++)
    for (idx[4] = idx[3]+1; idx[4] < count;     idx[4]++) {
        Card combo[5];
        for (int i = 0; i < 5; i++) combo[i] = cards[idx[i]];

        HandResult r = hand_eval_five(combo);
        if (!have_best || hand_compare(&r, &best) > 0) {
            best = r;
            have_best = 1;
        }
    }

    return best;
}
